### PR TITLE
Updated code for SS4

### PR DESCRIPTION
**Pull request by adrian-stein**
I should have got everything but can amend if anything is missed or not up to standard.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -1,20 +1,20 @@
...
-  "type": "silverstripe-module",
-  "description": "A module that indicates when people are editing the same page in the CMS",
-  "license": "BSD-3-Clause",
-  "keywords": ["silverstripe", "multi-user", "concurrent", "editing", "alert", "user", "content", "author"],
-  "authors": [
-    {
-      "name": "Julian Seidenberg",
-      "homepage": "http://www.deltaflow.com"
+    "name": "silverstripe/multiuser-editing-alert",
+    "type": "silverstripe-module",
```
Comment (dhensby) [inline composer.json]: vendormodule?

File: composer.json [hunk 2]
```diff
@@ -1,20 +1,20 @@
...
+    "license": "BSD-3-Clause",
+    "keywords": ["silverstripe", "multi-user", "concurrent", "editing", "alert", "user", "content", "author"],
+    "authors": [
+        {
+            "name": "Julian Seidenberg",
+            "homepage": "http://www.deltaflow.com"
+        }
+    ],
+    "require": {
+        "silverstripe/cms": "^4.0@dev"
```
Comment (dhensby) [inline composer.json]: why explicitly dev version?

File: src/MultiUserEditingController.php [hunk 3]
```diff
@@ -1,41 +1,51 @@
-<?php 
+<?php
+
+namespace MultiUserEditing;
```
Comment (dhensby) [inline src/MultiUserEditingController.php]: `SilverStripe\MultiUserEditing`

**Review by dhensby — CHANGES_REQUESTED** [review]
a few initial thoughts

**Comment by adrian-stein** [comment]
Il make the changes in a bit. How do I reference requirements once the module has been moved into the silverstripe vendor folder? The below doesn't work

`SilverStripe\Admin\LeftAndMain:`
`  extra_requirements_javascript:`
`    - 'vendor/silverstripe/multiuser-editing/js/multiUserEditing.js'`

**Comment by adrian-stein** [comment]
@dhensby I have made your changes

File: _config/multiuser.yml [hunk 4]
```diff
@@ -2,15 +2,20 @@
...
-Director:
+SilverStripe\Control\Director:
   rules:
-    'admin/editing//$Action/$ID': 'MultiUserEditingController'
-LeftAndMain:
+    'multiuser/editing//$Action/$ID': 'Silverstripe\MultiUserEditing\MultiUserEditingController'
+SilverStripe\Admin\LeftAndMain:
   extra_requirements_javascript:
-    - 'multiuser-editing/javascript/multiUserEditing.js'
+    - 'vendor/silverstripe/multiuser-editing/js/multiUserEditing.js'
```
Comment (dhensby) [inline _config/multiuser.yml]: `silverstripe/multiuser-editing:js/multiUserEditing.js`

File: _config/multiuser.yml [hunk 5]
```diff
@@ -2,15 +2,20 @@
...
   extra_requirements_css:
-    - 'multiuser-editing/css/multiuserediting.min.css'
-MultiUserEditingController:
+    - 'vendor/silverstripe/multiuser-editing/css/multiuserediting.min.css'
```
Comment (dhensby) [inline _config/multiuser.yml]: `silverstripe/multiuser-editing:css/multiuserediting.min.css`

File: js/multiUserEditing.js [hunk 6]
```diff
@@ -90,7 +90,7 @@ jQuery.entwine("multiUserEditing", function($) {
 				usersBR += "</ul>";
 
 				var dotHTML = '<div class="user-label multi-user-icon" title="' + users +
-					'"><img height="5" width="5" ' + 'src="multiuser-editing/images/' + dot + '"></div>';
+					'"><img height="5" width="5" ' + 'src="resources/silverstripe/multiuser-editing/images/' + dot + '"></div>';
```
Comment (dhensby) [inline js/multiUserEditing.js:93]: These hard-coded paths could be an issue - but I think we should leave them for now as I'm not sure the best way to resolve these paths in JS

**Review by dhensby — CHANGES_REQUESTED** [review]
just some vendormodule tweaks needed

**Comment by dhensby** [comment]
I'm happy with this once the last feedback is implemented.

@adrian-stein can you please do  a final test locally to make sure you're happy and let me know it's all working as expected?

Thanks

File: composer.json [hunk 7]
```diff
@@ -1,20 +1,26 @@
...
+        "homepage": "http://www.deltaflow.com"
+        }
+    ],
+    "require": {
+        "silverstripe/cms": "^4.0",
+        "silverstripe/vendor-plugin": "^1.0"
+    },
+    "minimum-stability": "dev",
+    "extra": {
+        "installer-name": "multiuser-editing",
```
Comment (dhensby) [inline composer.json]: you can remove the `installer-name` - please also add a branch alias for `dev-master` so it's `2.x-dev`, like so:

```json
"branch-alias": {
  "dev-master": "2.x-dev"
},
```

File: _config/multiuser.yml (hunk 5 above)
Comment (adrian-stein) [reply _config/multiuser.yml]: @dhensby  when I do that it comes up with [Emergency] Uncaught InvalidArgumentException: Can't find module 'silverstripe/multiuser-editing'. Is there something I am missing?

**Comment by adrian-stein** [comment]
@dhensby did you see my comment on _config/multiuser.yml? Any ideas on what I need to do?

File: _config/multiuser.yml (hunk 5 above)
Comment (dhensby) [reply _config/multiuser.yml]: sorry, it should be `silverstripe/multiuser-editing-alert:...`

File: _config/multiuser.yml (hunk 4 above)
Comment (dhensby) [reply _config/multiuser.yml]: `silverstripe/multiuser-editing-alert:...` sorry

**Comment by adrian-stein** [comment]
Thanks. Done. Also works as expected.

**Review by dhensby — APPROVED** [review]

**Comment by dhensby** [comment]
OK - I'm happy, thanks @adrian-stein 

@robbieaverill do you want to have a quick eye-over this and merge if you're happy too?

**Review by robbieaverill — CHANGES_REQUESTED** [review]
I haven't tested this, and it's a shame there aren't any unit tests, but it looks good overall. Only comment, could you please add an `.upgrade.yml` file map for the controller?

**Comment by adrian-stein** [comment]
That ok @robbieaverill @dhensby ?

**Comment by robbieaverill** [comment]
Nice, thank you!

**Pull request merged**